src/mlpack/core/data/scaler_methods/whitening.hpp

```diff
@@ -0,0 +1,147 @@
+/**
+ * @file whitening.hpp
+ * @author Jeffin Sam
+ *
+ * Whitening scaling to scale features, Using PCA and ZCA Whitening.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_DATA_WHITENING_SCALE_HPP
+#define MLPACK_CORE_DATA_WHITENING_SCALE_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/math/lin_alg.hpp>
+
+namespace mlpack {
+namespace data {
+
+/**
+ * A simple Whitening class.
+ *
+ * Whitens a matrix using the eigendecomposition of the covariance matrix.
+ * Whitening means the covariance matrix of the result is the identity matrix.
+ * 
+ * For whitening related formula and more info, check the link below.
+ * http://ufldl.stanford.edu/tutorial/unsupervised/PCAWhitening/
+ *
+ * @code
+ * arma::mat input;
+ * Load("train.csv", input);
+ * arma::mat output;
+ *
+ * // Scale the features using PCA.
+ * Whitening scale;
+ * scale.PCA(input, output);
+ *
+ * // Scale the features using ZCA.
+ * scale.ZCA(input, ouput);
+ * @endcode
+ */
+class Whitening
+{
+ public:
+  /**
+  * A constructor to set the regulatization parameter.
+  *
+  * @param eps Regularization parameter.
+  */
+  Whitening(double eps = 0.00005)
+  {
+    epsilon = eps;
+  }
+
+  /**
+  * Function for PCA whitening.
+  *
+  * @param input Dataset to scale features.
+  * @param output Output matrix with whitened features.
+  */
+  template<typename MatType>
+  void PCA(const MatType& input, MatType& output)
+  {
+    output.copy_size(input);
+    itemMean = arma::mean(input, 1);
+    // Centering the Dataset
+    output = (input.each_col() - itemMean);
+    arma::mat diag;
+    // Get eigenvectors and eigenvalues of covariance of input matrix.
+    eig_sym(eigenValues, eigenVectors, mlpack::math::ColumnCovariance(output));
+    // Generate diagonal matrix using 1 / sqrt(eigenvalues + epsilon) for each 
+    // value.
+    for (size_t i = 0; i < eigenValues.n_elem; i++)
+        eigenValues(i) = eigenValues(i) + epsilon;
```

Comment:
Looks like this doesn't take the square root of the inverse of the eigenvalues.  You could also do that as an Armadillo expression...

```
eigenValues = 1.0 / (arma::sqrt(eigenvalues) + epsilon);
```